File: src/decoder_re/BTMDecoder.java
package decoder;

import com.MiTAC.TRA.ATP.Tools.Byte2Number;
import com.MiTAC.TRA.ATP.decoder.waySidePacket.WaySideTelegramPacketDecoder;
import java.util.Vector;

/**
 * BTMDecoder is responsible for reassembling BTM (Balise Transmission Module) telegrams
 * from 5 data fragments received over MVB (Multifunction Vehicle Bus).
 * 
 * <p>Each complete BTM telegram is split into 5 fragments that arrive separately.
 * This class tracks multiple telegram sequences simultaneously and reassembles them.</p>
 * 
 * <h3>Data Structure:</h3>
 * <ul>
 *   <li><b>pointer</b>: Array tracking which sequence numbers are being collected (index = slot, value = sequence number)</li>
 *   <li><b>pnt</b>: 2D array tracking which fragments have been received (pnt[slot][fragmentNo] = 1 if received)</li>
 *   <li><b>btmData</b>: 3D array storing the actual fragment data (btmData[slot][fragmentNo][bytes])</li>
 * </ul>
 * 
 * <h3>Reassembly Process:</h3>
 * <ol>
 *   <li>When a fragment arrives (via setData), extract its sequence number from byte[0]</li>
 *   <li>Find or allocate a slot in the pointer array for this sequence number</li>
 *   <li>Store the fragment data in btmData[slot][fragmentNo]</li>
 *   <li>Mark the fragment as received in pnt[slot][fragmentNo] = 1</li>
 *   <li>When all 5 fragments are received (all pnt[slot][0-4] == 1), decode the complete telegram</li>
 *   <li>Reset the slot for reuse with the next sequence</li>
 * </ol>
 * 
 * <h3>Fragment Layout:</h3>
 * <pre>
 * Fragment 0: bytes 22-25 (4 bytes) -> output bytes 0-3
 * Fragment 1: bytes 1-25  (25 bytes) -> output bytes 4-28
 * Fragment 2: bytes 1-25  (25 bytes) -> output bytes 29-53
 * Fragment 3: bytes 1-25  (25 bytes) -> output bytes 54-78
 * Fragment 4: bytes 1-25  (25 bytes) -> output bytes 79-103
 * Total: 104 bytes
 * </pre>
 */
public class BTMDecoder {
  /**
   * Array of 10 slots to track up to 10 concurrent telegram sequences.
   * Each element stores the sequence number being collected in that slot,
   * or -1 if the slot is available.
   */
  int[] pointer = new int[10];
  
  /**
   * 2D array tracking which of the 5 fragments have been received for each slot.
   * pnt[slot][fragmentNo] = 1 when that fragment has been received.
   * pnt[slot][fragmentNo] = 0 when that fragment is still missing.
   */
  int[][] pnt = new int[10][5];
  
  /**
   * 3D array storing the actual fragment data.
   * btmData[slot][fragmentNo][bytes] contains up to 32 bytes per fragment.
   */
  byte[][][] btmData = new byte[10][5][32];
  
  private WaySideTelegramPacketDecoder wtpd;
  
  private boolean haveResult = false;
  
  public BTMDecoder() {
    init();
  }
  
  private void init() {
    for (int x = 0; x < this.pointer.length; x++) {
      this.pointer[x] = -1;
      for (int y = 0; y < (this.pnt[x]).length; y++)
        this.pnt[x][y] = 0; 
    } 
  }
  
  /**
   * Receives a BTM telegram fragment and stores it for reassembly.
   * 
   * @param bdata the fragment data (32 bytes), where byte[0] contains the sequence number
   * @param telegramNo the fragment number (1-5)
   */
  public void setData(byte[] bdata, int telegramNo) {
    int sequenceNumber = Byte2Number.getUnsigned(bdata[0]);
    int point = -1;
    int x;
    for (x = 0; x < this.pointer.length; x++) {
      if (this.pointer[x] == sequenceNumber)
        point = x; 
    } 
    if (point == -1)
      for (x = 0; x < this.pointer.length; x++) {
        if (this.pointer[x] == -1) {
          point = x;
          this.pointer[x] = sequenceNumber;
          break;
        } 
      }  
    if (point != -1)
      setData(bdata, point, telegramNo); 
  }
  
  private void setData(byte[] bdata, int point, int telegramNo) {
    this.haveResult = false;
    telegramNo--;
    this.pnt[point][telegramNo] = 1;
    this.btmData[point][telegramNo] = bdata;
    boolean collect5 = true;
    int x;
    for (x = 0; x < (this.pnt[point]).length; x++) {
      if (this.pnt[point][x] == 0)
        collect5 = false; 
    } 
    if (collect5) {
      this.haveResult = true;
      this.pointer[point] = -1;
      for (x = 0; x < (this.pnt[point]).length; x++)
        this.pnt[point][x] = 0; 
      decode(this.btmData[point]);
    } 
  }
  
  /**
   * Decodes a complete BTM telegram by extracting and concatenating data from all 5 fragments.
   * 
   * <p>The 5 fragments are reassembled into a single 104-byte array according to the
   * ERTMS/ETCS specification for BTM telegrams:</p>
   * <ul>
   *   <li>Fragment 0 (bytes 22-25): 4 bytes of header data</li>
   *   <li>Fragments 1-4 (bytes 1-25 each): 100 bytes of telegram payload</li>
   * </ul>
   * 
   * @param bdata 2D array containing the 5 fragments [fragment][bytes]
   */
  private void decode(byte[][] bdata) {
    try {
      byte[] data = new byte[104];
      for (int x = 0; x < bdata.length; x++) {
        int y;
        byte[] tmp = bdata[x];
        int loc = -1;
        switch (x) {
          case 0:
            for (y = 0; y < 4; y++)
              data[y] = tmp[22 + y]; 
            break;
          case 1:
            loc = (loc < 0) ? 4 : loc;
          case 2:
            loc = (loc < 0) ? 29 : loc;
          case 3:
            loc = (loc < 0) ? 54 : loc;
          case 4:
            loc = (loc < 0) ? 79 : loc;
            for (y = 0; y < 25; y++)
              data[loc + y] = tmp[1 + y]; 
            break;
        } 
      } 
      if (this.wtpd == null) {
        this.wtpd = new WaySideTelegramPacketDecoder(data);
      } else {
        this.wtpd.set(data);
        this.wtpd.go();
      } 
    } catch (Exception ex) {
      ex.printStackTrace();
    } 
  }
  
  public String toString() {
    return this.wtpd.getTelegramStream();
  }
  
  public int[] getBTMData() {
    return this.wtpd.get();
  }
  
  public int getNID_BG() {
    return this.wtpd.getNID_BG();
  }
  
  public Vector getGradient() {
    int[][] gradient = this.wtpd.getGradient();
    Vector rtn = new Vector();
    for (int x = 0; x < gradient.length; x++) {
      int[] zz = gradient[x];
      rtn.add(new Integer(zz[0]));
      rtn.add(new Integer(zz[1]));
      rtn.add(new Integer(zz[2]));
    } 
    while (rtn.size() < 27)
      rtn.add(new Integer(-1)); 
    return rtn;
  }
  
  public Vector getCurve() {
    int[][] curve = this.wtpd.getCurve();
    Vector rtn = new Vector();
    for (int x = 0; x < curve.length; x++) {
      int[] zz = curve[x];
      rtn.add(new Integer(zz[0]));
      rtn.add(new Integer(zz[1]));
      rtn.add(new Integer(zz[2]));
    } 
    while (rtn.size() < 27)
      rtn.add(new Integer(-1)); 
    return rtn;
  }
  
  public int[] getDetail() {
    return new int[] { this.wtpd.getNID_BG(), this.wtpd.getMCOUNT() };
  }
  
  public boolean isDataCollectFinished() {
    return this.haveResult;
  }
}

File: src/atp_re/decoders/btm_decoder.py
"""
BTM (Balise Transmission Module) Decoder with fragment reassembly.

This module provides functionality to decode and reassemble BTM telegrams
from 5 fragments into complete 104-byte telegrams, matching the Java
BTMDecoder functionality.
"""

from dataclasses import dataclass
from typing import Optional, List, Tuple
from .byte_utils import Byte2Number


@dataclass
class BTMFragment:
    """
    Represents a single BTM telegram fragment.
    
    Attributes:
        sequence_number: Telegram sequence number (0-255)
        telegram_number: Fragment number within sequence (1-5)
        data: Raw fragment data
    """
    sequence_number: int
    telegram_number: int
    data: bytes
    
    def to_dict(self):
        """Convert to dictionary for display."""
        return {
            "sequence_number": self.sequence_number,
            "telegram_number": self.telegram_number,
            "data_length": len(self.data),
            "data_hex": self.data.hex(),
        }


@dataclass
class BTMTelegram:
    """
    Represents a complete reassembled BTM telegram.
    
    Attributes:
        sequence_number: Telegram sequence number
        data: Complete 104-byte telegram data
        nid_bg: Balise group identifier
        m_count: Message count
    """
    sequence_number: int
    data: bytes
    nid_bg: Optional[int] = None
    m_count: Optional[int] = None
    
    def to_dict(self):
        """Convert to dictionary for display."""
        return {
            "sequence_number": self.sequence_number,
            "data_length": len(self.data),
            "data_hex": self.data.hex(),
            "nid_bg": self.nid_bg,
            "m_count": self.m_count,
        }


class BTMDecoder:
    """
    Decoder for BTM telegrams with fragment reassembly.
    
    BTM telegrams are transmitted in 5 fragments:
    - Fragment 1: Contains bytes 22-25 (4 bytes header)
    - Fragment 2: Contains bytes 1-25 at position 4-28
    - Fragment 3: Contains bytes 1-25 at position 29-53
    - Fragment 4: Contains bytes 1-25 at position 54-78
    - Fragment 5: Contains bytes 1-25 at position 79-103
    
    Total reassembled telegram: 104 bytes
    
    The decoder maintains 10 slots for tracking different sequence numbers
    simultaneously, allowing parallel reassembly of multiple telegrams.
    """
    
    NUM_SLOTS = 10
    NUM_FRAGMENTS = 5
    TELEGRAM_SIZE = 104
    
    def __init__(self):
        """Initialize BTM decoder with empty slots."""
        # Track sequence number for each slot (-1 means empty)
        self._sequence_slots: List[int] = [-1] * self.NUM_SLOTS
        
        # Track which fragments have been received for each slot
        # 0 = not received, 1 = received
        self._fragment_status: List[List[int]] = [
            [0] * self.NUM_FRAGMENTS for _ in range(self.NUM_SLOTS)
        ]
        
        # Store fragment data for each slot
        self._fragment_data: List[List[Optional[bytes]]] = [
            [None] * self.NUM_FRAGMENTS for _ in range(self.NUM_SLOTS)
        ]
        
        # Last completed telegram
        self._last_telegram: Optional[BTMTelegram] = None
    
    def add_fragment(self, fragment_data: bytes, telegram_number: int) -> Optional[BTMTelegram]:
        """
        Add a BTM fragment and attempt reassembly.
        
        Args:
            fragment_data: Raw fragment data (first byte is sequence number)
            telegram_number: Fragment number (1-5)
            
        Returns:
            BTMTelegram if reassembly is complete, None otherwise
            
        Raises:
            ValueError: If telegram_number is invalid or data is too short
        """
        if telegram_number < 1 or telegram_number > self.NUM_FRAGMENTS:
            raise ValueError(
                f"Invalid telegram number: {telegram_number} "
                f"(must be 1-{self.NUM_FRAGMENTS})"
            )
        
        if len(fragment_data) < 1:
            raise ValueError("Fragment data too short")
        
        # Extract sequence number from first byte
        sequence_number = Byte2Number.get_unsigned(fragment_data[0])
        
        # Find or allocate a slot for this sequence number
        slot_idx = self._find_or_allocate_slot(sequence_number)
        
        if slot_idx is None:
            # No available slots (all 10 are in use)
            # This shouldn't happen in normal operation
            return None
        
        # Store fragment (telegram_number is 1-based, convert to 0-based index)
        fragment_idx = telegram_number - 1
        self._fragment_status[slot_idx][fragment_idx] = 1
        self._fragment_data[slot_idx][fragment_idx] = fragment_data
        
        # Check if all 5 fragments have been received
        if all(status == 1 for status in self._fragment_status[slot_idx]):
            # Reassemble telegram
            telegram = self._reassemble_telegram(slot_idx, sequence_number)
            
            # Clear slot for reuse
            self._clear_slot(slot_idx)
            
            self._last_telegram = telegram
            return telegram
        
        return None
    
    def _find_or_allocate_slot(self, sequence_number: int) -> Optional[int]:
        """
        Find existing slot for sequence number or allocate a new one.
        
        Args:
            sequence_number: Sequence number to find or allocate
            
        Returns:
            Slot index (0-9) or None if no slots available
        """
        # First, check if sequence number already has a slot
        for i in range(self.NUM_SLOTS):
            if self._sequence_slots[i] == sequence_number:
                return i
        
        # Find an empty slot
        for i in range(self.NUM_SLOTS):
            if self._sequence_slots[i] == -1:
                self._sequence_slots[i] = sequence_number
                return i
        
        # No available slots
        return None
    
    def _clear_slot(self, slot_idx: int) -> None:
        """
        Clear a slot for reuse.
        
        Args:
            slot_idx: Slot index to clear
        """
        self._sequence_slots[slot_idx] = -1
        self._fragment_status[slot_idx] = [0] * self.NUM_FRAGMENTS
        self._fragment_data[slot_idx] = [None] * self.NUM_FRAGMENTS
    
    def _reassemble_telegram(self, slot_idx: int, sequence_number: int) -> BTMTelegram:
        """
        Reassemble complete telegram from fragments.
        
        Args:
            slot_idx: Slot index containing fragments
            sequence_number: Sequence number of telegram
            
        Returns:
            Complete BTMTelegram
        """
        # Allocate 104-byte telegram buffer
        telegram_data = bytearray(self.TELEGRAM_SIZE)
        
        # Reassemble according to BTM fragment structure
        for frag_idx in range(self.NUM_FRAGMENTS):
            fragment = self._fragment_data[slot_idx][frag_idx]
            if fragment is None:
                continue
            
            if frag_idx == 0:
                # Fragment 1: bytes 22-25 (4 bytes) -> positions 0-3
                if len(fragment) >= 26:
                    telegram_data[0:4] = fragment[22:26]
            elif frag_idx == 1:
                # Fragment 2: bytes 1-25 (25 bytes) -> positions 4-28
                if len(fragment) >= 26:
                    telegram_data[4:29] = fragment[1:26]
            elif frag_idx == 2:
                # Fragment 3: bytes 1-25 (25 bytes) -> positions 29-53
                if len(fragment) >= 26:
                    telegram_data[29:54] = fragment[1:26]
            elif frag_idx == 3:
                # Fragment 4: bytes 1-25 (25 bytes) -> positions 54-78
                if len(fragment) >= 26:
                    telegram_data[54:79] = fragment[1:26]
            elif frag_idx == 4:
                # Fragment 5: bytes 1-25 (25 bytes) -> positions 79-103
                if len(fragment) >= 26:
                    telegram_data[79:104] = fragment[1:26]
        
        # Create telegram object
        telegram = BTMTelegram(
            sequence_number=sequence_number,
            data=bytes(telegram_data)
        )
        
        # TODO: Parse NID_BG and M_COUNT from telegram data
        # This would require wayside packet decoder implementation
        
        return telegram
    
    def get_last_telegram(self) -> Optional[BTMTelegram]:
        """
        Get the last successfully reassembled telegram.
        
        Returns:
            Last BTMTelegram or None if no telegram has been reassembled
        """
        return self._last_telegram
    
    def reset(self) -> None:
        """Reset decoder state, clearing all slots."""
        self._sequence_slots = [-1] * self.NUM_SLOTS
        self._fragment_status = [
            [0] * self.NUM_FRAGMENTS for _ in range(self.NUM_SLOTS)
        ]
        self._fragment_data = [
            [None] * self.NUM_FRAGMENTS for _ in range(self.NUM_SLOTS)
        ]
        self._last_telegram = None
    
    def get_pending_sequences(self) -> List[Tuple[int, int]]:
        """
        Get list of pending (incomplete) telegram sequences.
        
        Returns:
            List of tuples (sequence_number, fragments_received)
        """
        pending = []
        for i in range(self.NUM_SLOTS):
            if self._sequence_slots[i] != -1:
                fragments_received = sum(self._fragment_status[i])
                pending.append((self._sequence_slots[i], fragments_received))
        return pending
